File: merge_sort.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <list>
#include <string>
#include <stdexcept>
#include <chrono>

using namespace std;
using namespace std::chrono;

struct RowData {
    int number;
    string text;

    RowData(int number, string text) {
        this->number = number;
        this->text = text;
    }
};

vector<RowData> readCSVRange(string& filename) {
    vector<RowData> list;
    int number, comma;
    string text, line;
    ifstream file(filename);
    if (!file.is_open()) {
        throw runtime_error("Error reading file: " + filename);
    }

    while (getline(file, line)) {
        comma = line.find(',');
        if (comma != -1) {
            number = stoi(line.substr(0, comma));
            text = line.substr(comma + 1);
            list.push_back(RowData(number, text));
        }
    }
    file.close();
    return list;
}

void writeStepsToFile(string& filename, vector<RowData>& data) {
    ofstream out(filename);
    if (!out.is_open()) {
        throw runtime_error("Error writing to file: " + filename);
    }

    for (int i = 0; i < data.size(); ++i) {
        out << data[i].number << "," << data[i].text << "\n";
    }
    out.close();
}

void merge(vector<RowData>& S, int left, int mid, int right) {
    list<RowData> L, R;

    for (int i = 0; i < mid - left + 1; ++i)
        L.push_back(S[left + i]);
    for (int j = 0; j < right - mid; ++j)
        R.push_back(S[mid + 1 + j]);

    int k = left;
    while (!L.empty() && !R.empty()) {
        if (L.front().number <= R.front().number) {
            S[k++] = L.front();
            L.pop_front();
        } else {
            S[k++] = R.front();
            R.pop_front();
        }
    }

    while (!L.empty()) {
        S[k++] = L.front();
        L.pop_front();
    }

    while (!R.empty()) {
        S[k++] = R.front();
        R.pop_front();
    }
}

void mergeSort(vector<RowData>& S, int left, int right) {
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSort(S, left, mid);
        mergeSort(S, mid + 1, right);
        merge(S, left, mid, right);
    }
}

int main() {
    string inputFile;
    cout << "Enter dataset filename: ";
    getline(cin, inputFile);

    vector<RowData> data = readCSVRange(inputFile);
    if (data.empty()) {
        throw runtime_error("Error: The dataset is empty or could not be read.");
    }

    string outputFile = "merge_sort_" + to_string(data.size()) + ".csv";

    auto start = high_resolution_clock::now();
    mergeSort(data, 0, data.size() - 1);
    auto end = high_resolution_clock::now();

    auto duration = duration_cast<seconds>(end - start);
    writeStepsToFile(outputFile, data);

    cout << "Merge sort steps written to " << outputFile << endl;
    cout << "Running time: " << duration.count() << " second" << endl;

    return 0;
}

File: merge_sort.java
import java.io.*;
import java.util.*;

public class merge_sort {

    public static class RowData {
        int number;
        String text;

        RowData(int number, String text) {
            this.number = number;
            this.text = text;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter dataset filename: ");
        String inputFile = scanner.nextLine();

        RowData[] numbers = readCSVRange(inputFile);
        if (numbers == null) {
            System.err.println("Error: Unable to read dataset.");
            scanner.close();
            return;
        }

        String outputFile = "merge_sort_" + numbers.length + ".csv";

        runAndTimeSort(numbers);

        writeStepsToFile(outputFile, numbers);
        System.out.println("Merge sort steps written to " + outputFile);

        scanner.close();
    }

    public static RowData[] readCSVRange(String filename) {
        List<RowData> list = new ArrayList<>();
        int number;
        String text;
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                // if (parts.length == 2) {
                number = Integer.parseInt(parts[0].trim());
                text = parts[1];
                list.add(new RowData(number, text));
                // }
            }
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
            return null;
        }
        return list.toArray(new RowData[0]);
    }

    public static void mergeSort(RowData[] S) {
        mergeSort(S, 0, S.length - 1);
    }

    public static void mergeSort(RowData[] S, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(S, left, mid);
            mergeSort(S, mid + 1, right);
            merge(S, left, mid, right);
        }
    }

    public static void merge(RowData[] S, int left, int mid, int right) {
        LinkedList<RowData> L = new LinkedList<>();
        for (int i = 0; i < mid - left + 1; i++)
            L.add(S[left + i]);

        LinkedList<RowData> R = new LinkedList<>();
        for (int j = 0; j < right - mid; j++)
            R.add(S[mid + 1 + j]);

        int k = left;

        while (!L.isEmpty() && !R.isEmpty()) {
            if (L.getFirst().number <= R.getFirst().number) {
                S[k++] = L.removeFirst();
            } else {
                S[k++] = R.removeFirst();
            }
        }

        while (!L.isEmpty()) {
            S[k++] = L.removeFirst();
        }

        while (!R.isEmpty()) {
            S[k++] = R.removeFirst();
        }
    }

    public static void writeStepsToFile(String filename, RowData[] data) {
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(filename))) {
            for (int i = 0; i < data.length; i++) {
                RowData row = data[i];
                bw.write(row.number + "," + row.text + "\n");
            }
        } catch (IOException e) {
            System.err.println("Error writing output file: " + e.getMessage());
        }
    }

    private static double runAndTimeSort(RowData[] data) {
        long start, end;
        double duration;
        start = System.currentTimeMillis();
        mergeSort(data);
        end = System.currentTimeMillis();
        duration = (end - start) / 1000;
        System.out.printf("Running time: %.3f second%n", duration);
        return duration;
    }
}